### PR TITLE
Persistent storage in queued_retry, backed by file storage extension

### PR DESCRIPTION
**Pull request by pmm-sumo**
**Description:**

Persistent queue implementation within queued_retry, aimed at being compatible with Jager's [BoundedQueue](https://github.com/jaegertracing/jaeger/blob/master/pkg/queue/bounded_queue.go) interface (providing a simple replacement) and backed by [file storage extension](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/storage/filestorage) for storing WAL.

Currently, to run the persistent queue, OpenTelemetry Collector Contrib with `enable_unstable` build tag is required.

**Link to tracking Issue:** #2285 

[Design doc](https://docs.google.com/document/d/1Y4vNthCGdYI61ezeAzL5dXWgiZ73y9eSjIDitk3zXsU/edit#)

**Testing:** Unit Tests and manual testing, more to come

**Documentation:** README.md updated, including an example

### CODE REVIEW

File: exporter/exporterhelper/queued_retry.go
Comment (djaglowski) [inline exporter/exporterhelper/queued_retry.go:183]: Based on the context, I think you should probably use `component.KindExporter`. (This only affects the name of the storage file.)

File: exporter/exporterhelper/logs.go
Comment (tigrannajaryan) [inline exporter/exporterhelper/logs.go:53]: Why do we need to unmarshal the context?

File: exporter/exporterhelper/wal.go
Comment (tigrannajaryan) [inline exporter/exporterhelper/wal.go:126]: Can we make the storage notify via channel when there is something to read instead of polling?

File: exporter/exporterhelper/wal.go
Comment (tigrannajaryan) [inline exporter/exporterhelper/wal.go:137]: Should this be changed to be more type safe? 
```suggestion
func (wq *WALQueue) Produce(item request) bool {
```

**Review by tigrannajaryan — COMMENTED** [review]
Reviewed partially.

File: exporter/exporterhelper/logs.go
Comment (pmm-sumo) [reply exporter/exporterhelper/logs.go:53]: I think we don't use context in exporters currently and it's indeed not needed. I thought about being able to use this code in other processors and the case when e.g. `k8sprocessor` would be consuming the output. Then, this adds good amount of complexity. Also, e.g. `batchprocessor` does not retain context either. I think it should be removed altogether and mentioned as a limitation

File: exporter/exporterhelper/wal.go
Comment (pmm-sumo) [reply exporter/exporterhelper/wal.go:126]: Good idea, I didn't like polling either, will take care of that

File: exporter/exporterhelper/queued_retry.go
Comment (pmm-sumo) [reply exporter/exporterhelper/queued_retry.go:77]: I was thinking about another setting for persistent data (since the number would be probably much different) but then maybe we should just keep the same setting to keep it simple

File: exporter/exporterhelper/README.md
Comment (pmm-sumo) [reply exporter/exporterhelper/README.md:35]: Yes, I believe we must do it. I didn't want to make this PR too complicated and taking consumers into play adds some level of complexity (actually, I am not sure what's the best implementation, so far, thought about separate set of keys for just the consumers to keep the currently processed items there)

**Comment by tigrannajaryan** [comment]
@pmm-sumo ping me when it is time to review again.

**Comment by pmm-sumo** [comment]
> @pmm-sumo ping me when it is time to review again.

Got a bit delayed but I am getting back to this now and should have an update by the end of the week

File: exporter/exporterhelper/wal.go
Comment (pmm-sumo) [reply exporter/exporterhelper/wal.go]: Actually, I kept it like that to match the BoundedQueue interface

**Comment by pmm-sumo** [comment]
@tigrannajaryan I have updated the codebase. It is now based on just the file storage interface (#3425). In effect, it requires OpenTelemetry Collector Contrib to run currently. Let me know if you would like to arrange it differently (and move most of the code here to Contrib - I wasn't sure how to do that without making it much more complex)

I renamed it to "Persistent Queue" and also addressed most of the comments. The major one missing is persistence for the consumers (i.e. when the `request` is taken from the queue). I was thinking about having separate items for each of the consumers backed by queue - that would require extra marshaling/unmarshaling but is probably the simplest/cleanest

File: exporter/exporterhelper/persistent_queue.go
Comment (pmm-sumo) [inline exporter/exporterhelper/persistent_queue.go:270]: Note, this currently deletes the item from file storage and can lead to data being lost (also, breaks one of the tests). I was thinking about addressing this by keeping the processed items separately (e.g. via one slot for each consumer) and doing some sort of confirmation callback from the consumer that the item has been processed

**Comment by tigrannajaryan** [comment]
> The major one missing is persistence for the consumers (i.e. when the `request` is taken from the queue).

@pmm-sumo sorry for a delayed review.

I think one typical approach to fix this problem is to have a queue with 3 pointers: write, read and delete. Reading does not delete the item from the queue. So the consuming end looks something like this:

```
for { 
  item := read(queue):
  err := send(item) // this is a blocking call that only returns when sending is done or fails.
  if !err {
    delete(queue, item)
  } else {
    // retry logic
  }
}
```

I think it is important to validate that the current storage interface allows to efficiently support such behavior.

**Comment by pmm-sumo** [comment]
> > The major one missing is persistence for the consumers (i.e. when the `request` is taken from the queue).
> 
> @pmm-sumo sorry for a delayed review.
> 
> I think one typical approach to fix this problem is to have a queue with 3 pointers: write, read and delete. Reading does not delete the item from the queue. So the consuming end looks something like this:
> (...)

Thank you @tigrannajaryan 

With multiple consumers this gets bit more complicated as the range between `delete` and `read` is not contiguous. I am finalising implementation right now that does use callbacks and another key/value which stores addresses of items that are currently being processed. Going to run a couple of tests and update the PR soon

**Comment by pmalek-sumo** [comment]
Hi @tigrannajaryan,

Can we get another review on this PR? 

@pmm-sumo is out for a couple of days now but we'll happily address any outstanding issues if there are any.

File: exporter/exporterhelper/persistent_storage.go
Comment (tigrannajaryan) [inline exporter/exporterhelper/persistent_storage.go:246]: Likewise here. This is 2 calls to Storage.Set, so would benefit from batching.

File: exporter/exporterhelper/persistent_storage.go
Comment (tigrannajaryan) [inline exporter/exporterhelper/persistent_storage.go:248]: And perhaps even this `Get` can be batched with the 2 previous `Set` calls (not sure if that provides any performance benefits though, worth checking).
If there are such benefits then we are not just looking for a simple batched `Set` but rather a transaction of Sets and Gets (similar to what bolt interface supports).
What do you think?

File: exporter/exporterhelper/persistent_storage.go
Comment (tigrannajaryan) [inline exporter/exporterhelper/persistent_storage.go:316]: This pair would also benefit from batching.

File: exporter/exporterhelper/README.md
Comment (tigrannajaryan) [inline exporter/exporterhelper/README.md:40]: @pmm-sumo I am a big fan of your diagrams!

File: exporter/exporterhelper/persistent_storage.go
Comment (pmm-sumo) [reply exporter/exporterhelper/persistent_storage.go:164]: Yes, that was exactly the intention

File: exporter/exporterhelper/persistent_storage.go
Comment (pmm-sumo) [reply exporter/exporterhelper/persistent_storage.go:167]: Oops, I think one of the loops is unnecessary @pmalek-sumo

File: exporter/exporterhelper/persistent_storage.go
Comment (pmm-sumo) [reply exporter/exporterhelper/persistent_storage.go:228]: This is a great idea. I think we could eventually also use some tools for garbage collection (essentially removing all unaccounted items). IIRC @djaglowski had some ideas when proposing the extension

File: exporter/exporterhelper/README.md
Comment (pmm-sumo) [reply exporter/exporterhelper/README.md:40]: Thank you! There is some complexity in how it works (which I was unable to simplify) so I thought the diagrams might be helpful in understanding what is going on here :)

File: exporter/exporterhelper/persistent_storage.go
Comment (pmalek-sumo) [reply exporter/exporterhelper/persistent_storage.go]: Done, thanks!

**Comment by pmalek-sumo** [comment]
Hi @tigrannajaryan,

Do we want to resolve the issue/challenge of batching in this PR or do we want to park this until later on? (perhaps create an issue for this)

If that's decided then we can think or resolving conflicts which got introduced in 0.29 if I understand the changelog correctly (and https://github.com/open-telemetry/opentelemetry-collector/pull/3455 seems to be the main culprit).

**Comment by tigrannajaryan** [comment]
> Do we want to resolve the issue/challenge of batching in this PR or do we want to park this until later on? (perhaps create an issue for this)

Can we discuss a few possible `Storage` interfaces that support batched Set or Get or even transactions with multiple operations mixed?

I would not want to merge `Storage` interface which we know is not good enough. It would be best to have the right interface from the beginning. This PR does not necessarily need to be refactored to get advantage of the improved `Storage`, it can be a separate PR, but it would be again great to prove that the improved `Storage` serves our needs.

**Comment by pmm-sumo** [comment]
@tigrannajaryan I have prepared an initial proposal in [contrib PR #4145](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/4145/files)

**Comment by tigrannajaryan** [comment]
@pmm-sumo do you want to keep this PR open or it is no longer needed?

**Comment by pmm-sumo** [comment]
> @pmm-sumo do you want to keep this PR open or it is no longer needed?

@tigrannajaryan it's still relevant. Now with the storage API merged I think even more so :)

**Comment by tigrannajaryan** [comment]
> > @pmm-sumo do you want to keep this PR open or it is no longer needed?
> 
> @tigrannajaryan it's still relevant. Now with the storage API merged I think even more so :)

Looking forward to an update that uses the new storage API (although if you prefer the use of batch operations may be a future improvement PR) :-)

**Comment by pmm-sumo** [comment]
> > > @pmm-sumo do you want to keep this PR open or it is no longer needed?
> > 
> > 
> > @tigrannajaryan it's still relevant. Now with the storage API merged I think even more so :)
> 
> Looking forward to an update that uses the new storage API (although if you prefer the use of batch operations may be a future improvement PR) :-)

I was keeping this PR updated with the batching changes and actually those are already incorporated :) I have used a [convenience layer](https://github.com/open-telemetry/opentelemetry-collector/pull/3274/files#diff-b9c787a9abc3a1e9495149bb3f0f11957bfd206459a0a22dfdd4c498f6a430de) (though I'm not sure if it's actually worth it)

**Comment by tigrannajaryan** [comment]
> I was keeping this PR updated with the batching changes and actually those are already incorporated :)

Nice. So this PR is ready for review?

**Comment by pmm-sumo** [comment]
> > I was keeping this PR updated with the batching changes and actually those are already incorporated :)
> 
> Nice. So this PR is ready for review?

Yes, I just rebased and resolved conflicts. Will do some tests later today or tomorrow to make sure nothing has broken since

File: exporter/exporterhelper/persistent_storage_batch.go
Comment (tigrannajaryan) [inline exporter/exporterhelper/persistent_storage_batch.go:213]: Nit: we use name `count` above and `size` here for the same thing.

**Comment by pmm-sumo** [comment]
Just a quick update, I started addressing the comments, should have an update early next week. Apologies for the delay

File: exporter/exporterhelper/persistent_queue.go
Comment (pmm-sumo) [reply exporter/exporterhelper/persistent_queue.go]: pq.storage has much more use and I decided to eventually remove persistentQueue one

File: exporter/exporterhelper/persistent_storage.go
Comment (pmm-sumo) [reply exporter/exporterhelper/persistent_storage.go]: Sure, used channel for messaging and got rid of the polling

File: exporter/exporterhelper/persistent_queue.go
Comment (pmm-sumo) [reply exporter/exporterhelper/persistent_queue.go:75]: There are two errors which might occur at `put` after refactoring:
* exceeding the capacity - currently logged at warning level: `pcs.logger.Warn("maximum queue capacity reached", zap.String(zapQueueNameKey, pcs.queueName))`
* failure when storing the item into the batch - logging causes of errors at debug level

I wanted to keep the interface of persistent queue compatible with BoundedQueue, hence just returning the bool. Now since the implementation of the latter was brought into OTC perhaps it could be refactored?

**Comment by pmm-sumo** [comment]
@tigrannajaryan I have addressed the comments (to keep track what was addressed and what was not, I click on "resolve" when applicable). There are few outstanding items - all related to BoundedQueue interface (and how PersistentQueue maps to it). I think that it is now copied into the Collector code, we might use it as an opportunity to refactor it, though since this PR has grew considerably - perhaps after merging this set of changes?

In a similar manner - I noticed that old, OC-based API is used for queued_retry metrics. I think it would be best to change it in a separate PR as well

File: exporter/exporterhelper/consumers_queue.go
Comment (pmm-sumo) [reply exporter/exporterhelper/consumers_queue.go:26]: I think that we could refactor BoundedQueue in a separate PR

File: exporter/exporterhelper/consumers_queue.go
Comment (pmm-sumo) [reply exporter/exporterhelper/consumers_queue.go:29]: Like above, I think we could organise it through a separate PR, now when BoundedQueue is copied into OpenTelemetry Collector codebase

**Comment by tigrannajaryan** [comment]
@pmm-sumo I am fine with doing remaining refactoring in future PRs.
The only concern I have with merging this right now is that there is RC work in progress and I want to make sure we do not break any existing functionality. How confident your are that if persistent queue config option is not enabled nothing changes in the behavior?
Should we perhaps hide the persistent queue functionality behind a feature flag (enablable at runtime or even only in a separate unstable exe build like we did previously for filelog).
@bogdandrutu what do you think? I am just back and don't know where we are with RC work. Can we take small risks and merge as is or we are close to GA release and should aim for stability?

**Comment by pmm-sumo** [comment]
That's a fair argument @tigrannajaryan 

> Should we perhaps hide the persistent queue functionality behind a feature flag (enablable at runtime or even only in a separate unstable exe build like we did previously for filelog).

If that helps, I can start looking into that and maybe separate the code a bit more at queued_retry (potentially at the expense of extra verbosity/duplicate code)

**Comment by tigrannajaryan** [comment]
> If that helps, I can start looking into that and maybe separate the code a bit more at queued_retry (potentially at the expense of extra verbosity/duplicate code)

Yes, I think it will be useful to have a clear decoupling of persistent queue implementation and if possible minimize the changes to any existing code. This will help to prove that the changes don't affect the existing behavior in any way.

File: exporter/exporterhelper/README.md
Comment (sumo-drosiek) [inline exporter/exporterhelper/README.md:24]:
```suggestion
  - `persistent_storage_enabled` (default = false): When set, enables persistence via a file storage extension (note, `enable_unstable` compilation flag needs to be enabled first)
```

File: exporter/exporterhelper/README.md
Comment (sumo-drosiek) [inline exporter/exporterhelper/README.md:37]:
```suggestion
It currently can be enabled only in OpenTelemetry Collector Contrib. Also, `enable_unstable` compilation flag needs to be enabled.
```

**Comment by pmm-sumo** [comment]
@tigrannajaryan I decoupled the code and introduced `enable_unstable` build flag (similarly as it was for the contrib). The decoupling could be brought further but it would require more code duplicated and I think what is left should be pretty much harmless.

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan I decoupled the code and introduced `enable_unstable` build flag (similarly as it was for the contrib). The decoupling could be brought further but it would require more code duplicated and I think what is left should be pretty much harmless.

@pmm-sumo Excellent. Please also add unstable build and test to Github actions (if not already).

File: exporter/exporterhelper/queued_retry_v1.go
Comment (tigrannajaryan) [inline exporter/exporterhelper/queued_retry_v1.go:1]: Would `queued_retry_inmemory.go` be a better name for this file?

File: exporter/exporterhelper/queued_retry_v2.go
Comment (tigrannajaryan) [inline exporter/exporterhelper/queued_retry_v2.go:1]: And this one perhaps `queued_retry_experimental.go`?

**Review by tigrannajaryan — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
@bogdandrutu PTAL before we merge this. Most of the code in this PR is disabled in regular build and is only effective when "unstable" build is created. We will enable the "unstable" part after we gain more confidence in it. This will also help prevent any disruption to our GA release plans.

File: exporter/exporterhelper/queued_retry_v1.go
Comment (pmm-sumo) [reply exporter/exporterhelper/queued_retry_v1.go]: Good idea, I renamed the file (and so did with `v2`)

**Comment by pmm-sumo** [comment]
> Please also add unstable build and test to Github actions (if not already).

I have created few more targets in `Makefile` and currently `make gotest` executes both `test` and `test-unstable`. As for the GitHub Actions, `build-and-test` creates also unstable set of binaries (though to be honest they are not very useful, one needs persistent storage implementation, which is available in Contrib)

**Comment by tigrannajaryan** [comment]
@bogdandrutu PTAL.

**Comment by tigrannajaryan** [comment]
@open-telemetry/collector-approvers if you reviewed this please approve if you are OK with the PR.

**Review by djaglowski — APPROVED** [review]
LGTM

**Comment by tigrannajaryan** [comment]
Since we have just made a release now is a good time to merge to have enough time till next release to fix any issues.
I am going merge the PR and if Bogdan or other approvers have comments they can be addressed in a future PR.

**Comment by tigrannajaryan** [comment]
@pmm-sumo thank you for your patience and the great effort to get this done!

**Pull request merged**
